rust.md

```diff
@@ -0,0 +1,146 @@
+* Setting `opt-level` to `'s'` tells the compiler to optimize for binary size rather than speed. `opt-level = 'z'` will optimize more aggressively for size, although at the cost of performance.
+
+Enable these options in the `release` profile by adding the following to `Cargo.toml`:
+```toml
+[profile.release]
+lto = true
+opt-level = 's'
+```
+
+#### Using [wasm-snip](https://github.com/rustwasm/wasm-snip)
+
+The Rust compiler, by default, includes panicking and formatting code in generated binaries. For the purpose of Ewasm contracts, this is useless.
+Using the `wasm-snip` tool, it is possible to replace the function bodies of such code with a single `unreachable` opcode, further shrinking the binary size.
+
+`wasm-snip` can be easily installed using `cargo`:
+```console
+$ cargo install wasm-snip
+```
+`wasm-snip` depends on the name section in the Wasm binary being present. Therefore, we must build with `debug` enabled. Add the following to the `[profile.release]` section in `Cargo.toml`:
+```toml
+debug = true
+```
+After build, run `wasm-snip` on the resulting binary. It should be located in `target/wasm32-unknown-unknown/release`:
+```console
+$ wasm-snip --snip-rust-panicking-code --snip-rust-fmt-code input.wasm -o output.wasm
+```
+
+## 3. Writing Ewasm contracts
+
+In addition to the normal constraints of the blockchain, the Ewasm [specification](https://github.com/ewasm/design) makes writing code for Ewasm contracts a little different from writing plain Rust.
+
+The following is a list of rules and best practices for writing Ewasm contracts.
+
+### `lib.rs`, not `main.rs`
+
+As stated earlier, using `main.rs` as a crate root will produce an executable rather than a library.
+
+### Export `main`
+
+Every Ewasm contract is required to export a `main` function as the main entry point for execution. In Rust, simply add the `pub` and `extern "C"` qualifiers to do this:
+```rust
+#[no_mangle]
+pub extern "C" fn main() {
+	// function body goes here
+}
+```
+The `no_mangle` directive should be included as well, to ensure that the compiler will not mangle the `main` identifier and invalidate it.
+
+### Use `finish` and `revert` to end execution
+
+When writing code for a native platform, ending execution properly is usually handled behind the scenes. In Ewasm, there are two EEI functions that signal to end execution:
+
+* [`finish`](https://github.com/ewasm/design/blob/master/eth_interface.md#finish) behaves like EVM's `return`. Simply put, it sets the output data and halts the VM.
+* [`revert`](https://github.com/ewasm/design/blob/master/eth_interface.md#revert) is the exact same thing as EVM's `revert`. It halts the VM and refunds any gas.
+
+### Minimize the use of the standard library
+
+Many constructs in Rust's standard library are not optimized for usage in Ewasm, or are outright unavailable due to dependence on a native syscall interface.
```

Comment:
Could you give a specific example of something that is unavailable or not optimized for use?

Reply:
Examples are below. As for something that is unavailable due to syscalls, most collections can simply use a custom allocator. So I might drop this line since nobody actually needs to read about that for this purpose.

Reply:
> I might drop this line since nobody actually needs to read about that for this purpose

Agree--where in doubt, keep it simple